### DISCUSSION TITLE
Considering alternative / better ways of naming function arguments

### DISCUSSION THREAD

**Discussion by levitte** [openssl/openssl#22317] [General Discussion]
#22296 was an issue that demonstrated a clash between argument names we use in OpenSSL public headers and other headers (in this case, the clash is with the name `gets`).  This is not the first time that we've had that happen, and it's potential if not likely that this will happen again.

It might be time to consider a standard for naming arguments, or not naming them at all (which is permitted in C standards), and to apply it broadly.

Possible variants, taking the example from #22296:

- Variant with no argument names at all (I personally dislike it):
  ``` C
  int BIO_meth_set_gets(BIO_METHOD *,
                        int (*) (BIO *, char *, int));
  ```
- Variant with names as comments:
  ``` C
  int BIO_meth_set_gets(BIO_METHOD * /* biom */,
                        int (* /* gets */) (BIO *, char *, int));
  ```
- Variant with names having a project prefix:
  ``` C
  int BIO_meth_set_gets(BIO_METHOD *ossl_biom,
                        int (*ossl_gets) (BIO *, char *, int));
  ```

This could ultimately lead to a [technical policy](https://github.com/openssl/technical-policies) along side the coding style, or an amendment of the coding style.

**Comment by levitte** [comment]
Ping @openssl/otc

**Comment by t8m** [comment]
My preference would be for the third option from above.

The question is where to apply it. To all arguments with names matching well-known libc functions? Or everywhere? Or just when we encounter a bug report like #22296? I'd vote for the 3rd or 1st option from these, against the second.

**Comment by levitte** [comment]
I'm thinking "everywhere" in the same way as we apply our function name prefix "everywhere".  In other words, primarily in new function declarations, but also allowing for refactoring for any reason.  Reasons to refactor could be quite broadly specified, including "I felt like it", but otherwise as response to all sort of findings.

Do note that an argument name change *isn't an API change*, 'cause it doesn't matter at all to the compiler (apart from clashes), which allows us to be quite frivolous with this sort of change.

**Comment by tom-cosgrove-arm** [comment]
Surely this is similar to "don't use reserved words as variable names"? It has happened before, and might happen again, but is it serious enough to require a prefix to all function names? In this instance, `gets` could easily be `gets_fn`, and might be more readable - I am concerned that there will be a negative impact to readability both in the implementation and the interface if this is done widely

**Reply by levitte** [reply]
Argument names, you mean...  `gets` is just an example and just so happens to be a function pointer in this particular case.

Essentially, that would constitute placing argument names we use into our "reserved namespace", which isn't uncommon.  For example, if you look at standard header files, you may see argument names prefixed with two underscores, which is the standard reserved namespace prefix for the C standard itself.

**Reply by t8m** [reply]
Yeah, `gets_fn` is more readable. And I do not think we should mandate argument names in prototypes being in our namespace.

**Reply by paulidale** [reply]
Honestly, a compiler that #defines `gets` is way way from the standard.  ISO C11 deprecates and removes `gets`.  We do target earlier C versions but it should be a function there.

**Comment by mspncp** [comment]
> Im not convinced that it is a good idea to rush in an ad-hoc fix for a compiler problem reported by a user #22296 based on suspicions only, without fully investigating the true cause, as suggested by @nhorman in [#22296 (comment)](https://github.com/openssl/openssl/issues/22296#issuecomment-1751350436).
>
> ...

see https://github.com/openssl/openssl/pull/22316#pullrequestreview-1666526875.

**Comment by tom-cosgrove-arm** [comment]
I am fine with not using `gets` as the name of a function argument. If it's been `#define`d somewhere, then that would also preclude using it as a variable name. But then, having a function argument or variable named `printf` wouldn't be great style, either. I'm sure there are compilers that will warn about shadowing global names.

But I don't see that this is a big enough or likely enough issue to warrant a change in the way function parameters are named.
